src/core/comm/coll.cc

```diff
@@ -0,0 +1,451 @@
+/* Copyright 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <atomic>
+#include <cstdlib>
+
+#ifndef LEGATE_USE_GASNET
+#include <stdint.h>
+#endif
+
+#include "coll.h"
+#include "legion.h"
+
+namespace legate {
+namespace comm {
+namespace coll {
+
+using namespace Legion;
+Logger log_coll("coll");
+
+#if defined(LEGATE_USE_GASNET)
+
+enum CollTag : int {
+  BCAST_TAG     = 0,
+  GATHER_TAG    = 1,
+  ALLTOALL_TAG  = 2,
+  ALLTOALLV_TAG = 3,
+  MAX_TAG       = 10,
+};
+
+static std::vector<MPI_Comm> mpi_comms;
+#else  // undef LEGATE_USE_GASNET
+static std::vector<ThreadComm> thread_comms;
+#endif
+
+static std::atomic<int> current_unique_id(0);
+
+static bool coll_inited = false;
+
+static constexpr int const MAX_NB_COMMS = 100;
```

Comment:
It will throw an error telling increasing the `MAX_NB_COMMS`. I am thinking about using `getenv` to make it user configurable, so it will be easier for users to increase the size. I explain the reason why we need this restriction in the next comment.